In [33]:
%reset

In [20]:
import pandas as pd
import importlib
from utils import price_opt_utils
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import StandardScaler


In [33]:
product = 'loader'
y_name = 'Standard Cost'
kwarg_product = dict(product = product)

In [34]:
X, y, relevant = price_opt_utils.get_x_y_relevant_dfs(**kwarg_product, y_name = y_name)

In [35]:
# model = LinearRegression().fit(X, y)

In [38]:
importlib.reload(price_opt_utils)

feature_selection_df_all = list()
# for selected_features in price_opt_utils.all_possible_feature_combinations(X):
for selected_features in price_opt_utils.iterative_selectiono_vif_ordered_features(X):
    
    X_selected = X[selected_features]
    
    # scaler = StandardScaler()
    # X_selected = pd.DataFrame(scaler.fit_transform(X_selected), columns = X_selected.columns)
    # X_selected = X_selected.apply(lambda x: x-x.mean())
    # y = y - y.mean()
    
    model = LinearRegression(fit_intercept = True).fit(X_selected, y)
    coefs = price_opt_utils.linear_regression_coefficients(model, X_selected.columns.tolist())
    
    formula = price_opt_utils.linear_regression_formula(model, X_selected.columns.tolist(), y_name = y_name)
    # print(formula)
    
    y_pred = model.predict(X_selected)
    results = price_opt_utils.regression_metrics(y, y_pred)
    # print(results)
    
    feature_selection_df_all.append(dict(selected_features = selected_features, 
                                          **results, formula = formula))
    
feature_selection_df_all = pd.DataFrame(feature_selection_df_all)

                 feature          VIF
0  Bucket volume std, m3   406.887220
1   Tramming Capacity, t  9040.580384
2    Weight Operating, t  7617.784568
3          Tilt force, t  3158.564302
4       Engine power, kW   387.784094
5          Lift force, t  2109.400746


KeyError: 1

In [37]:
feature_selection_df_all.loc[feature_selection_df_all['mape'].idxmin()]['formula']

'Standard Cost = 27995 + 22569 Tramming Capacity, t - 5019 Weight Operating, t - 55 Engine power, kW + 1112 Lift force,'

#### Standard Cost = -115020 - 1867 Hole diameter min, mm + 4015 Hole diameter max, mm + 4093 Flushing air, m3/min - 4082 Percussion power std, 

#### Standard Cost = 27995 + 22569 Tramming Capacity, t - 5019 Weight Operating, t - 55 Engine power, kW + 1112 Lift force,